plugin/src/node-builder.ts

```diff
@@ -93,35 +92,51 @@ const processorMap: ProcessorMap = {
 
 async function buildFromId(
   obj: Record<string, any>,
-  getFactory: (remoteType: string) => (node: IdentifiableRecord) => NodeInput,
+  getFactory: (remoteType: string) => (node: BulkResult) => NodeInput,
   gatsbyApi: SourceNodesArgs,
   options: ShopifyPluginOptions
 ) {
-  const [shopifyId, remoteType] = obj.id.match(pattern) || [];
+  const [, remoteType] = obj.id.match(pattern) || [];
 
   const Node = getFactory(remoteType);
   const processor = processorMap[remoteType] || (() => Promise.resolve());
 
   attachParentId(obj);
-  const node = Node({ ...obj, id: shopifyId });
+  const node = Node(obj);
   await processor(node, gatsbyApi, options);
 
   return node;
 }
 
 export function nodeBuilder(
-  nodeHelpers: NodeHelpers,
   gatsbyApi: SourceNodesArgs,
   options: ShopifyPluginOptions
 ): NodeBuilder {
   const factoryMap: {
-    [k: string]: (node: IdentifiableRecord) => NodeInput;
+    [k: string]: (node: BulkResult) => NodeInput;
   } = {};
+  function createNodeFactory(remoteType: string) {
+    return (result: BulkResult) => {
+      const id = result.id as string;
+      if (!id) {
+        throw new Error("Cannot build node without an ID");
+      }
+
+      return {
+        ...result,
+        shopifyId: id,
+        id: gatsbyApi.createNodeId(id),
+        internal: {
+          type: `Shopify${remoteType}`,
+          contentDigest: gatsbyApi.createContentDigest(result),
+        },
+      };
+    };
+  }
```

Comment:
My question here would be, why have a nodeFactory at all if the only thing you specify is remoteType? What's the benefit?

Reply:
It would be nice if you could put it in the backlog.

This function will give extra cognitive overhead in the future. When I read the code and encounter getFactory, it will make me need to look into createNodeFactory to only find out that it's not doing anything specific to the remoteType field.